python/tank/util/shotgun.py

```diff
@@ -224,24 +224,61 @@ def download_url(sg, url, location):
     :param location: path on disk where the payload should be written.
                      this path needs to exists and the current user needs
                      to have write permissions
+    :param bool use_url_extension: Optionally replace the file extension of the
+                                   input `location` value to match the extension
+                                   of the downloaded URL if it can be determined
+                                   and doesn't match already. If set to False, the
+                                   contents of `url` will be downloaded to `location`
+                                   without modification, preserving prior behavior.
+                                   Set this value to True if the file extension for the
+                                   URL to be downloaded cannot be determined in advance,
+                                   for example downloading the content of
+                                   https://my-site.shotgunstudio.com/thumbnail/full/Asset/1227.
+    :returns: Full filepath to the downloaded file. This may have been altered from
+              the input `location` if `use_url_extension` is True and the url's extension
+              was different than `location`.
     :raises: :class:`TankError` on failure.
     """
-    # grab proxy server settings from the shotgun API
-    if sg.config.proxy_handler:
+    # We only need to set the auth cookie for downloads from Shotgun server,
+    # input URLs like: https://my-site.shotgunstudio.com/thumbnail/full/Asset/1227
+    if sg.config.server in url:
+        # this method also handles proxy server settings from the shotgun API
+        __setup_sg_auth_and_proxy(sg)
+    elif sg.config.proxy_handler:
+        # These input URLs have generally already been authenticated and are
+        # in the form: https://sg-media-staging-usor-01.s3.amazonaws.com/9d93f...
+        # %3D&response-content-disposition=filename%3D%22jackpot_icon.png%22.
+        # Grab proxy server settings from the shotgun API
         opener = urllib2.build_opener(sg.config.proxy_handler)
+
         urllib2.install_opener(opener)
     
     # inherit the timeout value from the sg API    
     timeout = sg.config.timeout_secs
     
     # download the given url
     try:
+        request = urllib2.Request(url)
+        if sg.config.server in url:
+            # Currently no public access to Shotgun API _user_agents, so
+            # use directly for now.
+            request.add_header("user-agent", "; ".join(sg._user_agents))
```

Comment:
what's the purpose of this? It's just for metrics right? I don't think we should include that in this method.